Proposal for alternative configuration the Vue root path

balage1551 opened the pull request:
Hello!

I've though about our solution on Vue `rootDirectory` and `resourcesJarClass`.
And I found it dangerous.

Let's look at this function:
```kotlin
    @JvmStatic fun rootDirectory(path: String, location: Location) {
        rootDirectory = if (location == Location.CLASSPATH) PathMaster.classpathPath(path) else Paths.get(path)
    }
```

As you can see, the classpath is evaluated by `PathMaster` when the value is set. `PathMaster` uses the currently value of `resourcesJarClass` to evaluate the path. Therefore the order of assigning values to both directory and jar class is relevant: you have to set first the class and then the directory for correct effect.

This is dangerous. My suggestion is to handle the two parameters together. For external path, the property setter way (`rootDirectory = Paths.get()` stay the same.
For classpath relative setting, I propose the following:

```kotlin
    @JvmStatic fun rootDirectory(path: String, resourcesJarClass: Class<*> = PathMaster::class.java) {
        this.resourcesJarClass = resourcesJarClass
        rootDirectory = PathMaster.classpathPath(path)
    }
```

and making the `resourcesJarClass` internal. Also I propose to remove the old `fun rootDirectory(path: String, location: Location)` (you are planning to publish a new major version, so this breaking-of-code type change would fit in well), but alternatively, you may keep the old function, but marking it @Deprecated and mentioning in its message that this is dangerous.

What do you think?

balage1551 commented:
I've created a convenient method but I think it could be replaced by the Paths.get()/Path.of() call instead.

```kotlin
    // For completeness, you may add this convenient version for String based path setup (But property based is better)
    @JvmStatic fun externalRootDirectory(path: String) { rootDirectory = Paths.get(path) }
```

tipsy commented:
I agree that we should address the issue about order. Your proposal is:

```kotlin
JavalinVue.rootDirectory("/vue") // classpath
JavalinVue.rootDirectory("/vue", PathMaster::class.java) // classpath in different jar

JavalinVue.rootDirectory = Paths.get("src/resources/vue") // for external
JavalinVue.externalRootDirectory("src/resources/vue") // alt external
```

I guess the API should be similar to how it is for static files, so I'm a bit hesitant to get rid of the `Location` 🤔

We could consider a builder, like

```kotlin
JavalinVue.rootDirectory("/path").classPath() // classpath
JavalinVue.rootDirectory("/path").classPath(PathMaster::class.java) // classpath in different jar

JavalinVue.rootDirectory("/path").external() // external
```

What do you think about that?

We could change the static files API to something similar then. like `staticFiles("/public").urlPrefix("/assets").external()`

balage1551 commented:
I like your proposal.  And I like builders (especially fluent builders like your proposal. That's more consistent in Java. (The same mistake I did on Friday: forget how it will appear in Java.)

However, the handling of path should be done differently, because as soon as we split the class and path setting into two functions, they could be used in any order: 

```kotlin
JavalinVue.rootDirectory("/path").classPath()
```

or 

```kotlin
JavalinVue.classPath(PathMaster::class.java).rootDirectory("/path")
```

Now when you set the rootDirectory, you immediately call the PathMaster to evaluate it. We could do one of the following:

Prolonged the evaluation of the path until first use. Save the raw path and call the `PathMaster` to resolve the path only when it is first used. This assumes these variables won't be changed after initialization. This is normally the case, but nothing prevents to change these static values anytime at the moment, so either the whole JavalinVue should be an initialization time builder. This would be a wise and logical step (I can't think of any real-life example for altering these values after initialization, but it would change too much in the API, so I would avoid it.

The other option is to save the raw value, and reevaluate it in `classpath()` or `external()` function. This leads to a little more redundant code, but a better option.

The third option is to still keeping them in one function, like my proposal.

Your proposal is prettier, but needs to store the Location and the directory temporally. That's cries for real builders, not just fluent setters.

I can do some prototype to see how each option would look like.

tipsy commented:
> Your proposal is prettier, but needs to store the Location and the directory temporally. That's cries for real builders, not just fluent setters.

We could throw a `.build()` or a `.set()` at the end there. 

It would be nice if you could include staticfiles when you make your proposal :)

tipsy commented:
@balage1551 I've had a look at the static-files API now, and I came up with this:

```kotlin
addStaticFiles {
    it.directory("/public") // default "/public"
    it.location(Location.CLASSPATH) // default "Location.CLASSPATH"
    it.urlPrefix("/my-prefix") // default "/"
    it.precompress(true) // default true?
    it.aliasCheck() // default null
    it.header(key, value) // no headers set by default
    it.header(key, value) // can be called multiple times
} 
```

We could do something similar for JavalinVue:

```
JavalinVue.configure {
    it.option1
    it.option2
}
```

Then we will evaluate it all at the end of `configure`.

balage1551 commented:
Yes, this is what I was thinking about.
Most (all?) of the settings of JavalinVue is set before Javalin is used and make no sense to change them later. So switching to builder patters seems better. Also, builder patterns allow checking dependent properties (like jarClass, location, directory) before build.

But why `it`? Why not `this` and the receiver pattern.

```kotlin
JavalinVue {
    // All the fields are `private set` now

    // Assumes the isDev state may be determined in configuration time, independently from context
    @JvmStatic fun configure(isDev: Boolean, op: JavalinVueBuilder.() -> Unit ) = JavalinVueBuilder(isDev).apply(op).build()
}

class JavalinVueBuilder internal constructor(val isDev: Boolean) {
    @JvmField var optimizeDependencies = true
    @JvmField var stateFunction: (Context) -> Any = { mapOf<String, String>() }
    @JvmField var cacheControl = "no-cache, no-store, must-revalidate"
    var rootDirectory: Path = PathMaster.defaultLocation(isDev)
        private set

    // Convenient methods
    fun externalSource(path: String) {
        rootDirectory = Paths.get(path)
    }

    fun externalSource(path: Path) {
        rootDirectory = path
    }

    fun internalSource(path: String, jarClass: Class<*> = PathMaster::class.java) {
        rootDirectory = when {
            jarClass.getResource(path).toURI().scheme == "jar" -> PathMaster.fileSystem.getPath(path) // we're inside a jar
            else -> Paths.get(jarClass.getResource(path).toURI()) // we're not in jar (probably running from IDE)
        }
    }

    internal fun build() {
        JavalinVue.rootDirectory = rootDirectory
        JavalinVue.optimizeDependencies = optimizeDependencies
        JavalinVue.cacheControl = cacheControl
        JavalinVue.stateFunction = stateFunction
        
        // cachedPaths and cachedDependencyResolver should be initialized here, too!
    }
}

```

This is just a skeleton for my idea and has to figure out some issues which may alter the code strongly:

Why is `isDev` evaluation is prolonged until the first use? It makes everything more complex and dangerous? Couldn't it be decided on configuration time? 
I see that it is now bound to the context, but as it is evaluated only once and the context is ignored all consecutive calls, it is not really based on the context at all. It is even misleading. 
Let's take the original settings, when `isDev` is determined whether the url is localhost or not. It hints the system could run in a hybrid mode where some of the urls are local (and handled as dev) and others are external and handled normally. But the url isn't taken into account after the first call.

So the big question: could the `isDev` flag be determined in configuration time (which would make the whole path handling cleaner and simpler) or should its evaluation be prolonged until runtime?

_Also, sorry for testing my versions in your CI by PR, but I have no maven installed (I use gradle) so I can't build the app locally._

balage1551 commented:
Another question is whether JavalinVue should be an `object` and being mutable, or if we use builder pattern, isn't it cleaner to have JavalinVue a global variable and the underlining class is immutable.

On use site there is no difference, but would avoid unintended alter of the state out of the configuration builder and makes it clean that that values are immutable. (If we use overload, we could even give the config function the same name as the variable making it even more invisible.)

```kotlin

class JavelinVueConfig internal constructor(val x : Int)

var JavelinVue : JavelinVueConfig = JavVueBuilder().build()
    internal set

@Suppress("FunctionName")
@JvmOverloads
fun JavelinVue(op: JavVueBuilder.() -> Unit = {}) {
    JavelinVue = JavVueBuilder().apply(op).build()
}

class JavVueBuilder internal constructor() {
    var x : Int = 1
    internal fun build(): JavelinVueConfig { return JavelinVueConfig(x) }
}

fun main() {
    println(JavelinVue.x)
    JavelinVue {
        x = 2
    }
    println(JavelinVue.x)
    // Not allowed: JavelinVue.x = 5
}
```

tipsy commented:
> But why `it`? Why not this and the receiver pattern.

This is how the other APIs in Javalin work. You have 

```java
.create(config -> {
    ...
    config.asyncRequestTimeout = ...
    config.autogenerateEtags = ...
    config.compressionStrategy(Brotli(4), Gzip(6)) 
    config.contextPath = ...
}
```

```java
.ws("/websocket/:path", ws -> {
    ws.onConnect(ctx -> ...);
});
```

```java
.events(event -> {
    event.serverStarting
    event.serverStarted
    event.serverStartFailed
    ...
}
```

These are all made using `Consumer<MyConfig>`  signatures, which work the same in both Kotlin and Java.

--- 

> Why is isDev evaluation is prolonged until the first use? It makes everything more complex and dangerous? Couldn't it be decided on configuration time?

It could be, this just make it work "out of the box" for the vast majority of users, so they don't have to think about configuring it (they will get instant reloads locally, and cached files in prod). I'm pretty happy with how this works, and we haven't had any comments or complaints on it yet, but I'm always open to suggestions.

---

> Also, sorry for testing my versions in your CI by PR, but I have no maven installed (I use gradle) so I can't build the app locally.

I don't mind that at all, but Javalin uses Maven Wrapper, so you can do `./mvnw test` locally and it will setup Maven and run the tests for you, you don't have to have Maven installed :)

balage1551 commented:
Now all is clear and understood. :-)

This also makes the object (static) pattern superior over singleton.

balage1551 commented:
As I had little experience in creating clean api in both languages, I made some experiments:

```kotlin

fun configIt(op: (AB) -> Unit) = AB().also(op).build()

fun configThis(op: AB.() -> Unit) = AB().apply(op).build()

fun configFI(op: Consumer<AB>) {
    val ab = AB()
    op.accept(ab)
    ab.build()
}

@JvmOverloads
fun configJvmOverload(op: AB.() -> Unit) = AB().apply(op).build()
@JvmOverloads
fun configJvmOverload(op: Consumer<AB>) = configJvmOverload { op.accept(this) }
```

Usage (Kotlin):

```kotlin
fun main() {

    // Kotlin: Boilercode (it)
    // Java: Boilercode (needs return null)
    configIt { it.x = 12 }

    // Kotlin: Clean
    // Java: Boilercode (needs return null)
    configThis { x = 13 }

    // Kotlin: Lots of boilercode
    // Java: Clean
    configFI(Consumer { it.x = 14 })

    // Kotlin: Clean
    // Java: Clean
    // API: Overload, so boilercode
    configJvmOverload {
        x = 15
    }
}
```

Usage (java):

```java
public class Jav {
    public static void main(String[] args) {
        // Boilercode
        configIt( c -> {
            c.x = 12;
            return null;
        });

        // Boilercode
        configThis( c -> {
            c.x = 13;
            return null;
        });

        // Clean
        configFI( c -> {
            c.x = 14;
        });

        // Clean
        configJvmOverload( c -> {
            c.x = 14;
        });
    }
}
```

tipsy commented:
I've found that writing the APIs in Java instead of Kotlib is the best option. You only write the (clean, clean, clean in your example).

balage1551 commented:
So what do you think something like this:

```kotlin

object JavalinVue {
    @JvmStatic
    var rootDirectory: Path? = null
        private set

    @JvmStatic
    var location: Location = Location.CLASSPATH
        private set

    @JvmStatic
    var jarClass: Class<*> = JavalinVue::class.java
        private set


    fun externalRoot(path: Path) {
        this.location = Location.EXTERNAL
        this.rootDirectory = path
    }

    fun externalRoot(path: String) {
        this.location = Location.EXTERNAL
        this.rootDirectory = Paths.get(path)
    }

    fun internalRoot(path: String = "/vue", jarClass: Class<*> = JavalinVue::class.java) {
        this.location = Location.CLASSPATH
        this.jarClass = jarClass
        this.rootDirectory = PathMaster.classpathPath(path)
    }

    @JvmStatic
    fun config(op: Consumer<JavalinVue>) {
        this{ op.accept(this) }
    }

    // Optional 1: makes Kotlin usage clean
    operator fun invoke(op: JavalinVue.() -> Unit) {
        this.config(Consumer { it.op() })
    }

    // Optional 2: makes Kotlin usage clean
    fun config(op: JavalinVue.() -> Unit) {
        this.config(Consumer { it.op() })
    }
}
```

Or do you mean to write the whole JavalinVue class in Java?

tipsy commented:
Write the class in Java. I left the public facing APIs in Java when I ported Javalin to Kotlin to get the cleanest codebase and public APIs.

balage1551 commented:
That's the info I've not known: that Javalin was originally Java. It fits so nicely to Kotlin, I thought it is a Kotlin lib with backward option for Java. :-D

balage1551 commented:
But I don't think `defaultLocation` is ok. 

As I think of it, it caused my original problem. ;-)
It assumes fatjar (for nondev) and no subproject (for dev). Both of it is false for me.

That's why I had to override the rootDirectory.

balage1551 commented:
Well I come up with a option which hopefully addresses all issues:

```kotlin

object JavalinVue {
    @JvmField var optimizeDependencies = true // only include required components for the route component
    @JvmField var cacheControl = "no-cache, no-store, must-revalidate"
    // [...]

    @JvmField
    var rootDirectory: Path? = null

    @JvmField
    var pathBuilder: PathBuilder = PathBuilder()

    @JvmStatic
    fun pathBuilder(op: Consumer<PathBuilder>) {
        op.accept(pathBuilder)
    }

    fun pathBuilder(op: PathBuilder.() -> Unit) {
        PathBuilder(op)
    }

    operator fun invoke(op : JavalinVue.() -> Unit) {
        this.apply(op)
    }
}

class PathBuilder {

    @JvmField
    var vueDir: String = "vue"

    @JvmField
    var location: Location = Location.CLASSPATH

    @JvmField
    var jarClass: Class<*> = PathBuilder::class.java

    @JvmField
    var projectDir: String = ""


    fun external() {
        this.location = Location.EXTERNAL
    }

    fun internal(jarClass: Class<*> = JavalinVue::class.java) {
        this.location = Location.CLASSPATH
        this.jarClass = jarClass
    }


    companion object {
        // Optional 1: makes Kotlin usage clean
        operator fun invoke(op: PathBuilder.() -> Unit): PathBuilder {
            return PathBuilder().apply(op)
        }

        @JvmStatic
        fun config(op: Consumer<PathBuilder>): PathBuilder {
            val pb = PathBuilder()
            op.accept(pb)
            return pb
        }
    }


    fun build(isDev: Boolean): Path {
        return when {
            // In dev mode, the IDE path is used
            isDev -> {
                Paths.get(projectDir, "src/main/resources/vue")
            }

            // Path is external
            location == Location.EXTERNAL -> {
                Paths.get(vueDir)
            }

            // Inside a jar, so use path within
            jarClass.getResource(vueDir).toURI().scheme == "jar" -> {
                val fileSystem by lazy {
                    FileSystems.newFileSystem(
                        jarClass.getResource("").toURI(),
                        emptyMap<String, Any>()
                    )
                }

                fileSystem.getPath(vueDir)
            }

            // Otherwise
            else -> {
                Paths.get(jarClass.getResource(vueDir).toURI())
            }
        }
    }
}
```

As now there are 4 parameters of the path, I raised it into a new builder. 
The `JavalinVue` stores the builder. The original line in `VueComponent`:

```kotlin
        rootDirectory = rootDirectory ?: PathMaster.defaultLocation(isDev)
```

should be replaced to:
```kotlin
        rootDirectory = rootDirectory ?: JavalinVue.pathBuilder.build()
```

This gives solution for handling external and internal paths, the non-fat-jar and the subproject problems. Also gives a nice API in kotlin:

```kotlin
  JavalinVue {
        optimizeDependencies = false
        pathBuilder {
            vueDir = "/mydir/vue"
            internal(PathBuilder::class.java)
        }
    }
```

and in Java, too:

```java
        JavalinVue.optimizeDependencies = false;
        JavalinVue.pathBuilder(pb -> {
            pb.vueDir = "/";
        });
```

What do you think? It is simple and serves all use cases.

tipsy commented:
> What do you think? It is simple and serves all use cases.

It's not bad, but I want all of Javalin's public API to be consistent, so if we're going `Consumer<Config>` I think it's better to write that class in Java, like it is in all the other APIs. I like the idea of splitting out the pathbuilder into one builder/function.

Could you give writing it in Java a try?

balage1551 commented:
I'll do, but I'm still working on it. (Writing code in Java after almost 2 years is as slow as it was when I've begun to use Kotlin. ;-) Pascal-notation vs C-notation and semicolons are the hardest to get back.)

tipsy commented:
Hmm... In Kotlin 1.4 they added SAM conversions: https://kotlinlang.org/docs/whatsnew14.html#sam-conversions-for-kotlin-interfaces. I need to look into this.

balage1551 commented:
> 
> 
> Hmm... In Kotlin 1.4 they added SAM conversions: https://kotlinlang.org/docs/whatsnew14.html#sam-conversions-for-kotlin-interfaces. I need to look into this.

Yes, and some more good features. :-)
I haven't mentioned because for verbose usage I was not sure what is the target Kotlin version you use.

tipsy commented:
Since we're moving to Javalin 4 we can definitely bump to Kotlin 1.4. I'm converting all the public Java APIs here: https://github.com/tipsy/javalin/pull/1226

balage1551 commented:
And you asked me to migrate the JavalinVue to Java from Kotlin? ;-)

BTW, I'm ready with the first version, but I don't like it, yet. So tomorrow evening I am going to make another option.

tipsy commented:
I'm sorry, I forgot that they had fixed the SAM-conversion issue 😅

balage1551 commented:
No problem. The main challenge was to get used to Java again. ;-)

So the question, that if I refactor my solution anyway, could I move back the PathMaker (I renamed it to PathFactory) to Kotlin? I'll use the SAM conversation and the Consumer pattern to make clean config API in both languages.

tipsy commented:
> So the question, that if I refactor my solution anyway, could I move back the PathMaker (I renamed it to PathFactory) to Kotlin? I'll use the SAM conversation and the Consumer pattern to make clean config API in both languages.

Yea, definitely.

tipsy commented:
I decided against converting all the public APIs to Kotlin (reason in https://github.com/tipsy/javalin/pull/1226), but you can still go ahead and write this file in Kotlin, as long as it's easy to use from both languages :)

tipsy commented:
@balage1551 still working on this? I just finished the new static files API, it ended up looking like this:

```kotlin
javalin.addStaticFiles { staticFiles ->
    staticFiles.hostedPath = "/url-prefix"
    staticFiles.directory = "/public"
    staticFiles.location = Location.CLASSPATH
    staticFiles.aliasCheck = ContextHandler.ApproveAliases()
    staticFiles.precompress = true
    staticFiles.headers = mapOf(Header.CACHE_CONTROL to "max-age=31622400")
}
```

tipsy commented:
@balage1551 I had a look at this and made the following config classes:

```kotlin
class VueDirConfig {
    fun externalPath(path: String) {
        JavalinVue.rootDirectory = Paths.get(path)
    }
    @JvmOverloads
    fun classpathPath(path: String, resourcesJarClass: Class<*> = PathMaster::class.java) {
        JavalinVue.resourcesJarClass = resourcesJarClass // used by line below ...
        JavalinVue.rootDirectory = PathMaster.classpathPath(path)
    }
    fun explicitPath(path: Path) {
        JavalinVue.rootDirectory = path
    }
}

class VueVersionConfig {
    fun vue2() {
        JavalinVue.vueVersion = VueVersion.VUE_2
        JavalinVue.vueAppName = "Vue"
    }
    fun vue3(appName: String) {
        JavalinVue.vueVersion = VueVersion.VUE_3
        JavalinVue.vueAppName = appName
    }
}
```

Usage is like this:

```kotlin
JavalinVue.vueVersion { it.vue2() }
JavalinVue.vueDirectory { it.classpathPath("/vue", TestJavalinVue::class.java) }
```

What do you think? It just keeps the global state, so the change is a bit smaller.

tipsy commented:
PR: https://github.com/tipsy/javalin/pull/1238

tipsy commented:
I'll close this for now, but feel free to create a new one if you have improvement suggestions :)
